grub-core/loader/i386/skinit.c

```diff
@@ -0,0 +1,101 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <grub/cpu/relocator.h>
+#include <grub/dl.h>
+#include <grub/i386/skinit.h>
+#include <grub/i386/tpm.h>
+#include <grub/mm.h>
+#include <grub/slr_table.h>
+#include <grub/types.h>
+
+static inline grub_uint16_t
+grub_skinit_get_sl_size (void)
+{
+  grub_uint16_t *ptr = grub_slaunch_module ();
+
+  if (ptr == NULL)
+    return 0;
+
+  return ptr[1];
+}
+
+grub_err_t
+grub_skinit_boot_prepare (struct grub_relocator *rel,
+                          struct grub_slaunch_params *slparams)
+{
+  grub_uint32_t *apic = (grub_uint32_t *)0xfee00300ULL;
+  grub_relocator_chunk_t ch;
+  grub_err_t err;
+  void *dce_mem;
+  grub_uint32_t dce_size_aligned;
+
+  err = grub_relocator_alloc_chunk_align (rel, &ch, 0x1000000,
+                                          0xffffffff - GRUB_SLAUNCH_TPM_EVT_LOG_SIZE,
+                                          GRUB_SLAUNCH_TPM_EVT_LOG_SIZE, GRUB_PAGE_SIZE,
+                                          GRUB_RELOCATOR_PREFERENCE_NONE, 1);
+
+  if (err != GRUB_ERR_NONE)
+    return grub_error (err, "cannot alloc memory for TPM event log");
+
+  slparams->tpm_evt_log_base = get_physical_target_address (ch);
+  slparams->tpm_evt_log_size = GRUB_SLAUNCH_TPM_EVT_LOG_SIZE;
+  grub_memset (get_virtual_current_address (ch), 0, slparams->tpm_evt_log_size);
+
+  grub_dprintf ("slaunch", "tpm_evt_log_base = %lx, tpm_evt_log_size = %x\n",
+                (unsigned long) slparams->tpm_evt_log_base,
+                (unsigned) slparams->tpm_evt_log_size);
+
+  /* Contrary to the TXT, on AMD we do not have vendor-provided blobs in
+   * reserved memory, we are using normal RAM */
+  err = grub_relocator_alloc_chunk_align (rel, &ch, 0,
+                                          0xffffffff - GRUB_SKINIT_SLB_SIZE,
+                                          GRUB_SKINIT_SLB_SIZE,
+                                          GRUB_SKINIT_SLB_ALIGN,
+                                          GRUB_RELOCATOR_PREFERENCE_LOW, 1);
+
+  if (err != GRUB_ERR_NONE)
+    return grub_error (err, "cannot alloc memory for SLB");
+
+  slparams->dce_base = get_physical_target_address (ch);
+  slparams->dce_size = grub_skinit_get_sl_size ();
+
+  dce_mem = get_virtual_current_address (ch);
+  grub_memset (dce_mem, 0, GRUB_SKINIT_SLB_SIZE);
+  grub_memcpy (dce_mem, grub_slaunch_module(), slparams->dce_size);
+
+  /* SLRT needs to be at least 4-byte aligned per specification */
+  dce_size_aligned = ALIGN_UP (slparams->dce_size, 4);
+  if (dce_size_aligned + GRUB_PAGE_SIZE > GRUB_SKINIT_SLB_SIZE)
+    return grub_error (GRUB_ERR_BUG, "not enough space for SLRT table");
+
+  slparams->slr_table_base = slparams->dce_base + dce_size_aligned;
+  slparams->slr_table_size = GRUB_PAGE_SIZE;
+  slparams->slr_table_mem = (grub_uint8_t *) dce_mem + dce_size_aligned;
```

Comment:
I'm currently reworking SLK header format, so fields with offsets will be moved around. Most likely this will be something like this:

```
GLOBAL(sl_header)
	.word	_entry           /* SL header SKL offset to code start */
	.word	_end_of_measured /* SL header SKL measured length */
	.word	skl_info         /* Offset to SKL info with UUID and version */
	.word	bootloader_data  /* Offset to SLRT passed by the bootloader */
ENDDATA(sl_header)
```

First two words are defined by AMD, so they can't be redefined (although what we use them for may). `skl_info` is used in sanity check after build, but other than that it can be reordered if needed. `bootloader_data` used to be in place of `_end_of_measured`, but because of that whole SLK was measured, including BSS and data that could as well be built by SKL at runtime. SINIT sends SLK to the TPM for measuring, which is slow, so having it measure tens of KB of zeroes is a waste of time.

Memory layout than I'm aiming for is roughly:

- `sl_header`
- `.text` with all the code, `_entry` is somewhere within (currently at the very beginning, but doesn't have to be)
- `.rodata` with constants
- `skl_info` - it used to be in separate section, but perhaps it can be included in `.rodata`
- `_end_of_measured` - everything above is measured, constant and read-only, everything below isn't
- `.bss`, cleared by SKL because it can't trust GRUB (so `grub_memset` isn't strictly necessary here, except for SLRT)
- `bootloader_data` - this is where SLRT will live
- free space (hopefully)
- stack - `%esp` after SKINIT points to SLB+64K. We used to point it to stack located in `.text` but no need to measure it. SLK will check if stack is big enough, accounting for SLRT size, because it can't trust GRUB.
- there is no `.data`, everything that is currently there will have to be moved either to `.rodata` or `.bss`

So to sum up:
- `grub_skinit_get_sl_size()` may use `_end_of_measured` - everything else is initialized by SKL anyway
- `dce_size` isn't used anywhere but here (I think), but perhaps it should be always set to 64K
- `dce_size_aligned` (probably renamed) should be based on `bootloader_data`, and on SKL side it will be accessed directly, without manual alignment - linker script can assure that it will be 4B-aligned
- check for `(dce_size_aligned + GRUB_PAGE_SIZE > GRUB_SKINIT_SLB_SIZE)` is sufficient - SKL may change in a way that will change its stack requirements, and it should be able to handle this without having to update GRUB code.

Reply:
Took this into account:
1. https://github.com/TrenchBoot/grub/compare/ad13737578f935bcff302407860c2c6da7f73dab..a2af951d4f61ac7a5ea819eba49ce979c544a5ff#diff-ee6e12de6eb52b720e2e6f3792ff7228232d905b280bef353649ae579b9a110eR29
2. https://github.com/TrenchBoot/grub/compare/ad13737578f935bcff302407860c2c6da7f73dab..a2af951d4f61ac7a5ea819eba49ce979c544a5ff#diff-ee6e12de6eb52b720e2e6f3792ff7228232d905b280bef353649ae579b9a110eL76

> * `dce_size` isn't used anywhere but here (I think), but perhaps it should be always set to 64K

It's used in SLRT and is an accurate size of ACM in case of TXT, so used `_end_of_measured` for now.